consensus/propagation/have_wants.go

```diff
@@ -63,58 +64,130 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts) {
 		return
 	}
 
-	reqLimit := 1
-
-	// if enough requests have been made for the parts, don't request them.
-	for _, partIndex := range hc.GetTrueIndices() {
-		reqs := blockProp.countRequests(height, round, partIndex)
-		if len(reqs) >= reqLimit {
-			hc.SetIndex(partIndex, false)
-			// mark the part as fully requested.
-			fullReqs.SetIndex(partIndex, true)
-		}
-		// don't request the part from this peer if we've already requested it
-		// from them.
-		for _, p := range reqs {
-			// p == peer means we have already requested the part from this peer.
-			if p == peer {
-				hc.SetIndex(partIndex, false)
-			}
+	for _, index := range hc.GetTrueIndices() {
+		select {
+		case <-blockProp.ctx.Done():
+		case p.requestChan <- request{height: height, round: round, index: uint32(index)}:
 		}
 	}
 
-	if hc.IsEmpty() {
-		return
+	blockProp.broadcastHaves(haves, peer, int(parts.Total()))
+}
+
+// perPeerPerBlockConcurrentRequestLimit returns the maximum number of requests for a given block by peer.
+// it helps avoid downloading the whole block from a single peer.
+func perPeerPerBlockConcurrentRequestLimit() int64 {
+	// TODO implement
+	return 300
+}
+
+func (blockProp *Reactor) wantsSendingRoutine(ps *PeerState) {
+	var want *proptypes.WantParts
+	batchRequestCount := int64(0)
```

Comment:
why is this batch logic needed at all? can't we just drain as much in the channel or go ahead and send it? especially if we do 

https://github.com/celestiaorg/celestia-core/pull/1773/files#r2060940556

Reply:
when batching the wants to be sent, we don't update the peer state count until we send them. Otherwise, we could add a part index, update the number of concurrent requests in the peer state, we move to a new height, we drop those previous wants, then we will also need to decrease the concurrent requests. Adding and subtracting seems more confusing than keeping track of the count. 

We could update this to use the `len(want.Pars.GetTrueIndices)`, but we will need to always check if the `want` is nil before accessing that which is too much I guess